nixos/nextcloud: Add option for using object storage as primary storage

mitchmindtree opened the pull request:
This allows to declaratively configure an S3 class object storage as the
primary storage for the nextcloud service. Previously, this could only
be achieved by manually editing the `config.php`.

I've started testing this today with my own digitalocean nextcloud
instance, which now points to my digitalocean S3-compatible "Space" and
all appears to be working smoothly.

My motivation for this change is my recent discovery of how much cheaper
some S3-compatible object storage options are compared to digitalocean's
"Volume" options.

Implementation follows the "Simple Storage Service" instructions here:

https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html

I have neglected to implement a submodule for the OpenStack Swift
object storage as I don't personally have a use case for it or a method
to test it, however the new `nextcloud.objectstore.s3` submodule should
act as a useful guide for anyone who does wish to implement it.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mitchmindtree committed:
nixos/nextcloud: Add option for using object storage as primary storage

This allows to declaratively configure an S3 class object storage as the
primary storage for the nextcloud service. Previously, this could only
be achieved by manually editing the `config.php`.

I've started testing this today with my own digitalocean nextcloud
instance, which now points to my digitalocean S3-compatible "Space" and
all appears to be working smoothly.

My motivation for this change is my recent discovery of how much cheaper
some S3-compatible object storage options are compared to digitalocean's
"Volume" options.

Implementation follows the "Simple Storage Service" instructions here:

https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html

I have neglected to implement a submodule for the OpenStack Swift
object storage as I don't personally have a use case for it or a method
to test it, however the new `nextcloud.objectstore.s3` submodule should
act as a useful guide for anyone who does wish to implement it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-configure-nextcloud-to-use-s3-object-storage/15180/3

aanderse commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -304,6 +304,124 @@ in {
+              The access key for the S3 bucket.
+            '';
+          };
+          secret = mkOption {
```
Please remove this option.

mitchmindtree replied:
Can do, it will also simplify things a little. Shall we also remove `adminpass` and `dbpass` in favour of their `*File` alternatives too?

aanderse replied:
Ideally yes. We should work toward that. A separate PR perhaps?

aanderse reviewed (changes requested):
Will defer to @Ma27 for review, except this one point.

mitchmindtree committed:
nixos/nextcloud: Remove `objectstore.s3.secret` option

We should discourage users from adding secrets in a way that allows for
them to end up in the globally readable `/nix/store`. Users should use
the `objectstore.s3.secretFile` option instead.

aanderse commented:
Thank you @mitchmindtree!

Ma27 commented:
Yeah, I have it on my todo list, but it's kinda overloaded, so I'll have to request some patience here :)

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -304,6 +304,113 @@ in {
+            default = null;
+            example = "example.com";
+            description = ''
+              Required for some non-Amazon implementations.
```
Does that mean, it's for services that are not from Amazon, but speak the "S3 protocol"? Anyways, please give some more context here.

mitchmindtree replied:
I'm unsure about this! My limited understanding is that in some implementations of "S3-compatible" storage a hostname is required, whereas apparently it is not on Amazon? I'd have no idea having avoided Amazon, but I know it is required for my digital ocean space at least.

mitchmindtree replied:
I should add that my comment here provides just about as much clarity as [the nextcloud docs](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html#simple-storage-service-s3) do.

Ma27 replied:
Yeah, makes sense.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -304,6 +304,113 @@ in {
+            '';
+          };
+          useSsl = mkOption {
+            type = types.nullOr types.bool;
```
We either use SSL or we don't, correct? :) 
(and are there any reasons for turning it off? It should be `true` at least).

mitchmindtree replied:
The difference between `null` and us deciding on a default value is that in the case the value is `null`, we don't add a `use_ssl` line to the config at all and in turn fall back to whatever Nextcloud's default happens to be.

That said I understand if you'd prefer we try to encourage the use of SSL - I'm happy to change the default to `true`.

Ma27 replied:
> we don't add a use_ssl line to the config at all and in turn fall back to whatever Nextcloud's default happens to be

Can't we just use the fallback that Nextcloud uses (i.e. Nextcloud's default is our default). Having `null` always makes a special-case mandatory which isn't really nice.

mitchmindtree replied:
Sure, it's not documented anywhere but [judging by the src](https://github.com/nextcloud/server/blob/e2116e2fb226890341c548e3f7d79c0ac63c1b06/lib/private/Files/ObjectStore/S3ConnectionTrait.php#L83) it looks like SSL is enabled by default.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -304,6 +304,113 @@ in {
+          };
+        };
+      in mkOption {
+        type = types.nullOr (types.submodule {
```
I don't see why this is needed. You can just do

``` nix
{
  objectstore = {
    enable = mkEnableOption "...";
    /* all the other objectstore parameters */
  };
}
```

if you can only specify a single object store.

mitchmindtree replied:
The reason why most config is behind `objectstore.s3.*` and not directly behind `objectstore.*` is because nextcloud supports multiple classes of object storage, each with its own set of unique configuration parameters. According to [their docs](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html) nextcloud currently supports using S3, Openstack Swift V1 and V2 classes of object storage as primary storage. The current structure allows to avoid breaking downstream configurations when someone inevitably decides to add support for Openstack Swift or some other future object storage class that nextcloud might support.

W.r.t. using `nullOr`, I was unsure if you could have a submodule default to `{}` in the case that it has non-default options! I'm happy to remove the `nullOr` indirection and use `{}` as a default instead, it would certainly clean things up a little, though I must admit it's a bit strange to me that nix allows having a submodule type instance that is missing fields that have no defaults. I guess a submodule is just an attrset under the hood though.

Ma27 replied:
Then do `objectstore.s3.enable = mkEnableOption "...";`, please. A submodule is absolutely not needed here so far and just causes unnecessary complexity. If this really becomes necessary, we can switch to `types.attrsOf (types.submodule {})` sooner or later or implement the other obejctstore things manually if someone's willing to do that.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -304,6 +304,113 @@ in {
+        type = types.nullOr (types.submodule {
+          options = {
+            s3 = mkOption {
+              type = types.submodule {
```
Same here.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -375,13 +482,16 @@ in {
         }
-        { assertion = ((acfg.adminpass != null || acfg.adminpassFile != null)
-            && !(acfg.adminpass != null && acfg.adminpassFile != null));
+        { assertion = (acfg.adminpass != null) != (acfg.adminpassFile != null);
```
Please do this in a separate commit. While the change itself seems correct, I'd prefer to always do it. That way maintainers don't get commits when bisecting for an issue with a lot of unrelated changes.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -375,13 +482,16 @@ in {
+        { assertion = acfg.objectstore == null
+            || (lists.count (v: v.enable) (attrsets.attrValues acfg.objectstore)) == 1;
+          message = "If using objectstore class as primary storage exactly one class can be enabled.";
+        }
```
With my suggestion from above we can drop this assertion entirely.

mitchmindtree replied:
This check will become useful in the case that options are added for the other nextcloud-supported classes of object storage as mentioned above, as according to the nextcloud docs a user should only be allowed to select a single object storage class to use as primary storage.

Ma27 replied:
This feels like premature optimization. As suggested above, `objectstore.s3.<option>` is good-enough for a tentative design, more things (including such an assertion) can be added later when we need it.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -441,11 +551,33 @@ in {
+            || (c.objectstore != null && c.objectstore.s3.enable);
+          objectstoreConfig = let
+            class = if c.objectstore.s3.enable then "S3" else "";
+            args = if c.objectstore.s3.enable then c.objectstore.s3.arguments else {};
```
Nix is lazily evaluated. As soon as you drop the unnecessary submodule above and work with a single `enable` option, you can use e.g. `s3.arguments` (which should default to `{}` then rather than `null`) here without this if-else. I guess the same applies to `class =`.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -441,11 +551,33 @@ in {
+              ${optionalString (args.port != null) "'port' => ${toString args.port},"}
+              ${optionalString (args.useSsl != null) "'use_ssl' => ${if args.useSsl then "true" else "false"},"}
+              ${optionalString (args.region != null) "'region' => '${args.region}',"}
+              'use_path_style' => ${if args.usePathStyle then "true" else "false"},
```
You can use `lib.boolToString` here.

Ma27 reviewed (changes requested):
Since you mentioned "non-Amazon" implementations, can we use such an implementation in a VM test? :)

mitchmindtree commented:
> Since you mentioned "non-Amazon" implementations, can we use such an implementation in a VM test? :)

Oooo I'm unsure if there is a self-hosted S3-compatible object storage type out there? I'm personally using digital ocean's implementation at the moment. I'd imagine that it *would* be possible to do this with the Openstack Swift class object storage once options are added for it, though I've no idea of what's involved in setting it up.

Ma27 commented:
Yeah fine. I guess I'll run some tests against an S3 bucket and if it works, we can merge as soon as we cleared up all the other points.

mitchmindtree committed:
nixos/nextcloud: Simplify `objectstore.s3` options, remove submodule

Removes the submodule in favour of using an attrset.

Also:

- Makes better use of nix's laziness in config expansion.
- Makes use of `boolToString` where applicable.

mitchmindtree committed:
nixos/nextcloud: Make `objectstore.s3.useSsl` explicitly true by default

This appears to match the nextcloud default behaviour observed here:

https://github.com/nextcloud/server/blob/e2116e2fb226890341c548e3f7d79c0ac63c1b06/lib/private/Files/ObjectStore/S3ConnectionTrait.php#L83

mitchmindtree committed:
nixos/nextcloud: Account for nix_read_secret refactor in exception msg

Previously, the `nix_read_pwd` function was only used for reading the
`dbpassFile`, however it has since been refactored to handle reading
other secret files too. This fixes the message of the exception that is
thrown in the case that the file is not present so that it no longer
refers specifically to the `dbpass` file.

mitchmindtree commented:
Okydoke @Ma27, I've rebased onto the latest master and added two more commits addressing your comments, along with another fixing the message of the exception that is thrown in the case that the specified secret file doesn't exist.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -479,14 +563,29 @@ in {
           c = cfg.config;
           writePhpArrary = a: "[${concatMapStringsSep "," (val: ''"${toString val}"'') a}]";
+          requiresReadSecretFunction = c.dbpassFile != null || c.objectstore.s3.enable;
+          objectstoreConfig = let s3 = c.objectstore.s3; in optionalString s3.enable '''objectstore' => [
```
PHP doesn't really care about indentation and this is not only weird to read, but also highly ambiguous because `'''` inside a multi-line string is actually `''` in the evaluated result.

Ma27 reviewed (commented):
Apart from a minor nitpick, the change is looking good now.

I'd like to test it against an S3 bucket (I'll get to it today or later this week), then we can merge. When I check it out locally, I can fix the styling-issue I pointed out by myself (assuming you allow maintainers to push to this branch).

mitchmindtree committed:
nixos/nextcloud: Fix ambiguity in `objectstoreConfig` string

Previously this was a little tricky to read and had the potential to
cause some ambiguity in string parsing.

mitchmindtree commented:
> assuming you allow maintainers to push to this branch

Is that what this refers to?

![Screenshot from 2021-10-05 17-12-55](https://user-images.githubusercontent.com/4587373/135976883-ffe28628-c2ad-4748-8f73-977304290724.png)

Either way, I've had a go at addressing the styling nit just in case.
